uro98/assignment3/FindAllWordsTest.java

```diff
@@ -0,0 +1,66 @@
+import static org.junit.Assert.*;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
+
+import org.junit.Test;
+
+public class FindAllWordsTest {
+
+  @Test
+  public void testFindAllWords() {
+    String[][] array = new String[][] {
+      {"A", "A", "R"},
+      {"T", "C", "D"}
+    };
+    
+    Grid grid = new Grid(2, 3);
+    grid.populate(array);
+    
+    Dictionary dictionary = new Dictionary();
+    dictionary.addWords(Arrays.asList("CAR", "CARD", "CART", "CAT"));
+    
+    Set<String> wordSet = FindAllWords.findAllWords(grid, dictionary);
+    
+    Set<String> expected = new HashSet<String>();
+    expected.add("CAR");
+    expected.add("CARD");
+    expected.add("CAT");
+    
+    assertEquals(expected, wordSet);
+    assertFalse(dictionary.getWordSet().equals(wordSet));
+  }
+
+  @Test
+  public void testSearchWords() {
```

Comment:
Not really needed. `searchWords` is an implementation detail and should be private, and tests typically use only public interface. There are rare cases when tests need to use normally hidden features of the class, but they are very rare.

Reply:
Are private methods usually not unit tested then? Why shouldn't they be?

Reply:
Private methods are usually not tested directly, only via public interface of the class. Reasoning: if public methods work correctly, then private methods that are used to implement them, work too. Of course then one has to make sure that test data set is sufficient to trigger all interesting paths in private methods.

Of course you can also do the trick that was done in Assignment 3 presentation (Laura should have shared it with you in "solutions" channel on Slack) - make those methods not private and not public, but package private. This lets you test them as long as test class is in the same package. Downside is that this breaks hermetization and exposes things that should normally be implementation detail.

There are also more advanced tricks like `@VisibleForTesting` annotation, which makes method visible in tests, but in other code it will be visible as private. Or approach that "if a method is complex enough to have dedicated set of tests, then maybe it should be part of separate class, where it will be public".

In general there is a balance on keeping interface of the class small and focused, and keeping it easily testable.

If you want to keep your tests for helper methods, feel free to keep them, just make the methods package private.

Reply:
I see, thank you for the detailed answer!